### PR TITLE
fix(42829) ignore preceeding jsx whitespace

### LINKED ISSUE #42829: React/JSX-based code actions don't ignore whitespace
*TS Template added by @mjbvz*

**TypeScript Version**: 4.3.0-dev.20210216

**Search Terms**
- code action
- quick fix
- jsx 

---

Issue Type: <b>Bug</b>

VS Code version: Code 1.52.0 (940b5f4bb5fa47866a54529ed759d95d09ee80be, 2020-12-10T22:43:50.597Z)
OS version: Linux x64 5.8.0-41-generic

For code actions that apply to blocks, leading and trailing whitespace is typically ignored:

Peek 2021-02-11 23-29

However, for code actions that apply to React/JSX blocks, this does not appear to be the case:

Peek 2021-02-11 23-24

Note in the above image that selecting the block does not appear to be sufficient; the selection has to be at the start of the JSX tag for the code actions to become available.

This seems like a pretty unfortunate usability and discoverability issue -- at first I wasn't certain whether JSX refactorings were completely broken!

---

### Expected behaviour

Code actions that apply to JSX within a selection will be displayed regardless of leading and trailing whitespace in the selection. This matches the behaviour of code actions that apply to non-JSX JavaScript and TypeScript code, including within a JSX/TSX file (for example, refactoring a block to a new function in module scope).

### Actual behaviour

Code actions for JSX are only shown when the beginning of the selection is at the start of the relevant block of JSX. Trailing whitespace seems to be permissible.

### To reproduce

I've created a minimal repro case here: https://github.com/lostfictions/vscode-jsx-actions-repro.

To reproduce the issue:

1. Clone the repo.
2. Enter the repo directory and run `npm install`.
3. Run `code --user-data-dir /tmp/code-user-tmp --extensions-dir /tmp/code-ext-tmp .` for a Code instance with a fresh profile and all extensions disabled.
4. Open `src/index.tsx`.
5. Select lines 4 to 10, including all leading whitespace. The code action icon will display, and if clicked will offer refactorings:
    image
6. Select lines 15 to 19, including all leading whitespace. The code action icon will not display:
    image
7. Select lines 15 to 19, but this time ensure the selection begins at line 15 at the first non-whitespace character, the `<` in `<div>`. This time the code action icon will display as normal and offer the expected refactorings:
    image

---

For convenience, here are the listings for `index.tsx`, `tsconfig.json`, and `package.json` in the above example:

`index.tsx`:

```tsx
import React from 'react';

export function NormalFunctionThatDoesNotExhibitIssue() {
  if(true) {
    for(let i=0;i<100;i++) {
      console.log('doing some work...', i)
    }
  }

  
}

export default function ComponentThatExhibitsIssue() {
  return <div>
    <div className="some-nested-data">
      hello from my nested component
    </div>

    
  </div>
}
```

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "target": "es2019",
    "module": "esnext",
    "esModuleInterop": true,
    "jsx": "preserve",
    "strict": true
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

`package.json`:

```json
{
  "name": "vscode-jsx-actions-repro",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@types/react": "^17.0.2",
    "@types/react-dom": "^17.0.1",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "typescript": "^4.1.5"
  }
}
```

### PR DESCRIPTION
**Pull request by ZakMiller**
Fixes #42829

### CODE REVIEW

**Comment by ZakMiller** [comment]
@jessetrinity is there something I can do to get this reviewed? First contribution

**Comment by jessetrinity** [comment]
@ZakMiller we try our best to prioritize issues based on their severity and our bandwidth. If an issue hasn't gotten attention in a long time, pinging a reviewer is fine. Thank you for your patience, I will take a look now! :)

File: src/services/utilities.ts [hunk 1]
```diff
@@ -1141,6 +1141,20 @@ namespace ts {
         }
     }
 
+    /**
+     * Returns the first token where position is in [start, end),
+     * excluding `JsxText` tokens containing only whitespace.
+     */
+    export function findFirstNonWhitespaceToken(sourceFile: SourceFile, position: number): Node | undefined {
```
Comment (jessetrinity) [inline src/services/utilities.ts:1148]:
```suggestion
    export function findFirstNonJsxWhitespaceToken(sourceFile: SourceFile, position: number): Node | undefined {

```

**Review by jessetrinity — APPROVED** [review]
Just one nitpick on the function name but otherwise looks great!

**Comment by jessetrinity** [comment]
Closing and reopening to hopefully get a fresh run.

**Pull request merged**